File: des--elev_rud_model_cpp_ws/src/ad_hb_msgs/src/ad_hb_msgs.cpp
#include <ros/ros.h>
#include <cmath>
#include <std_msgs/Float64.h>

class HeartbeatPublisher {
private:
    ros::NodeHandle nh_;
    ros::Timer publish_timer_;
    ros::Publisher psi_ref_pub_;
    ros::Publisher z_ref_pub_;
    ros::Publisher n_rpm_pub_;

public:
    HeartbeatPublisher() : nh_("~") {
        // Create publishers with desired topic names
        psi_ref_pub_ = nh_.advertise<std_msgs::Float64>("/psi_ref_msg", 10);
        z_ref_pub_ = nh_.advertise<std_msgs::Float64>("/z_ref_msg", 10);
        n_rpm_pub_ = nh_.advertise<std_msgs::Float64>("/n_rpm_msg", 10);

        // Create a timer to publish the values periodically
        publish_timer_ = nh_.createTimer(ros::Duration(1.0), &HeartbeatPublisher::publishValues, this);
    }

    void publishValues(const ros::TimerEvent& event) {
        // Assign values to n_rpm
        double n_rpm = 1500;

        // Calculate the current time in seconds
        double currentTime = ros::Time::now().toSec();

        // Calculate the desired depth using a sine function
        double period = 20;
        double amplitude = 2.0;
        double frequency = 1.0 / period;
        double z_ref = 250 + amplitude * sin(2 * M_PI * frequency * currentTime);

        // Calculate the desired heading angle (psi_ref) based on the desired trajectory
        double psi_ref = atan2(-amplitude * frequency * 2 * M_PI * cos(2 * M_PI * frequency * currentTime), 1);

        std_msgs::Float64 psi_ref_msg;
        psi_ref_msg.data = psi_ref;
        psi_ref_pub_.publish(psi_ref_msg);

        std_msgs::Float64 z_ref_msg;
        z_ref_msg.data = z_ref;
        z_ref_pub_.publish(z_ref_msg);

        std_msgs::Float64 n_rpm_msg;
        n_rpm_msg.data = n_rpm;
        n_rpm_pub_.publish(n_rpm_msg);
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "ad_hb_msgs");

    HeartbeatPublisher heartbeatPublisher;

    ros::spin();

    return 0;
}

File: des--elev_rud_model_TTpy_ws/src/ad_hb_msgs/nodes/ad_hb_msgs.py
#!/usr/bin/env python3

import rospy
import math
from std_msgs.msg import Float64
from avl_bsd_translator.msg import HeartbeatMsg

def heartbeat_callback(msg):
    # Assign values to n_rpm
    n_rpm = 1500

    # Calculate the current time in seconds
    current_time = rospy.get_time()

    # Calculate the desired depth using a sine function
    period = 20
    amplitude = 1.0
    frequency = 1.0 / period
    z_ref = 250 + amplitude * math.sin(2 * math.pi * frequency * current_time)

    # Calculate the desired heading angle (psi_ref) based on the desired trajectory
    psi_ref = math.atan2(-amplitude * frequency * 2 * math.pi * math.cos(2 * math.pi * frequency * current_time), 1)

    # Set x_des and y_des to create straight line trajectories
    x_des = 5.0  # Replace 5.0 with the desired x-coordinate value for the trajectory
    y_des = -3.0  # Replace -3.0 with the desired y-coordinate value for the trajectory

    # Publish psi_ref, z_ref, x_des, y_des, and n_rpm values to their respective topics
    psi_ref_msg = Float64()
    psi_ref_msg.data = psi_ref
    psi_ref_pub.publish(psi_ref_msg)

    z_ref_msg = Float64()
    z_ref_msg.data = z_ref
    z_ref_pub.publish(z_ref_msg)

    x_des_msg = Float64()
    x_des_msg.data = x_des
    x_des_pub.publish(x_des_msg)

    y_des_msg = Float64()
    y_des_msg.data = y_des
    y_des_pub.publish(y_des_msg)

    n_rpm_msg = Float64()
    n_rpm_msg.data = n_rpm
    n_rpm_pub.publish(n_rpm_msg)

    print("we are here:")
if __name__ == '__main__':
    rospy.init_node('ad_hb_msgs')

    # Create publishers for psi_ref, z_ref, x_des, y_des, and n_rpm values
    psi_ref_pub = rospy.Publisher('psi_ref_msg', Float64, queue_size=10)
    z_ref_pub = rospy.Publisher('z_ref_msg', Float64, queue_size=10)
    x_des_pub = rospy.Publisher('x_des_msg', Float64, queue_size=10)
    y_des_pub = rospy.Publisher('y_des_msg', Float64, queue_size=10)
    n_rpm_pub = rospy.Publisher('n_rpm_msg', Float64, queue_size=10)

    # Subscribe to the heartbeat message
    rospy.Subscriber('/bsd/heartbeat', HeartbeatMsg, heartbeat_callback)

    rospy.spin()
